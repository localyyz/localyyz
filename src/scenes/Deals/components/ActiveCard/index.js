import React from "react";
import { View, StyleSheet, Text, TouchableOpacity } from "react-native";

// third party
import { Provider, observer, inject } from "mobx-react/native";

// custom
import { ConstrainedAspectImage, ProgressBar } from "localyyz/components";
import { Colours, Sizes, Styles } from "localyyz/constants";

// custom
import ActiveDealUIStore from "./store";

@inject(stores => ({
  dealStore: stores.dealStore
}))
@observer
export default class ActiveCard extends React.Component {
  constructor(props) {
    super(props);
    this.store = new ActiveDealUIStore(this.props);
  }

  get product() {
    return this.store.products[0] || {};
  }

  render() {
    return this.store.products.length > 0 ? (
      <Provider activeDealStore={this.store}>
        <View style={styles.container}>
          <View style={styles.card}>
            <Text style={styles.title}>{this.props.name}</Text>
            <Text style={styles.text}>{this.props.description}</Text>
          </View>
          <View style={styles.activeContent}>
            <ConstrainedAspectImage
              source={{ uri: this.product.imageUrl }}
              constrainHeight={Sizes.Height / 4}/>
            <TouchableOpacity
              onPress={() =>
                this.props.navigation.navigate("Product", {
                  product: this.product,
                  dealStore: this.props.dealStore,
                  activeDealStore: this.store,
                  dealId: this.props.id
                })
              }>
              <View style={styles.button}>
                <Text style={styles.buttonLabel}>
                  {`Buy now $${this.product.price}`}
                </Text>
              </View>
            </TouchableOpacity>
            <View style={styles.social}>
              <Text style={styles.socialLabel}>
                {`${
                  this.store.usersViewing
                    ? `${this.store.usersViewing} currently viewing â€” `
                    : ""
                }${this.store.quantityAvailable} available`}
              </Text>
              <ProgressBar progress={this.store.progress} />
            </View>
          </View>
        </View>
      </Provider>
    ) : null;
  }
}

const styles = StyleSheet.create({
  container: {
    marginBottom: Sizes.InnerFrame / 2,
    backgroundColor: Colours.Foreground
  },

  card: {
    padding: Sizes.InnerFrame
  },

  activeContent: {
    alignItems: "center"
  },

  social: {
    width: "100%",
    alignItems: "center",
    justifyContent: "center"
  },

  socialLabel: {
    ...Styles.Text,
    ...Styles.Emphasized,
    ...Styles.SmallText,
    color: Colours.Primary,
    marginBottom: Sizes.InnerFrame
  },

  button: {
    ...Styles.RoundedButton,
    margin: Sizes.InnerFrame
  },

  buttonLabel: {
    ...Styles.Text,
    ...Styles.Emphasized,
    ...Styles.Alternate
  },

  title: {
    ...Styles.Text,
    ...Styles.Title,
    marginBottom: Sizes.InnerFrame / 2
  },

  text: {
    ...Styles.Text
  }
});
