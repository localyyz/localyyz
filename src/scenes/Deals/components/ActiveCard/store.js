import { Animated, Easing } from "react-native";

// third party
import { observable, runInAction, computed } from "mobx";

// custom
import { ApiInstance } from "localyyz/global";
import { Deal } from "localyyz/models";

// constants
const DEBUG = true;
const REFRESH_INTERVAL = 5000;

export default class ActiveDealUIStore {
  @observable deal;
  @observable iteration = 0;

  constructor(deal) {
    // data
    this.deal = deal;
    this.progress = new Animated.Value(0);

    // bindings
    this.updateProgress = this.updateProgress.bind(this);
    this.refresh = this.refresh.bind(this);

    // initial data load
    this.updateProgress(this.deal.percentageClaimed);
    this.refresh();
  }

  updateProgress(percentage) {
    Animated.timing(this.progress, {
      toValue: percentage,
      duration: REFRESH_INTERVAL / 5,
      easing: Easing.out(Easing.ease)
    }).start();
  }

  refresh(interval = REFRESH_INTERVAL) {
    this.fetch().then(() => {
      this._refreshTimeout && clearTimeout(this._refreshTimeout);

      // fire first, then wait
      this._refreshTimeout = setTimeout(() => this.refresh(interval), interval);
    });
  }

  async fetch() {
    let response = await ApiInstance.get(this.deal.origin);
    if (response && response.status < 400 && response.data) {
      // inject origin data from previous deal data
      let deal = new Deal({ ...response.data, origin: this.deal.origin });
      !this.deal.equals(deal)
        && runInAction("[ACTION] Fetching Today's deal", () => {
          this.deal = deal;
          this.iteration = this.iteration + 1;
          DEBUG
            && console.log(
              "[ActiveDeal] Deal updated:",
              deal,
              "from",
              this.deal.origin
            );

          // update progress
          this.updateProgress(this.deal.percentageClaimed);
        });
    }
  }

  @computed
  get products() {
    return this.deal.products || [];
  }
}
