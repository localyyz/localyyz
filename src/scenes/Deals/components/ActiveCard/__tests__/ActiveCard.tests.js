import React from "react";
import Renderer from "react-test-renderer";

// local
import ActiveCard from "../index";

const Component = ActiveCard;

// constants
const NAME = "Deals/ActiveCard";
const PROPS = {
  deal: {
    name: "Tes Store",
    description: "Test Description",
    products: [
      {
        brand: "Test",
        title: "Test Title",
        associatedPhotos: [{ imageUrl: "www.localyyz.com" }],
        price: 6.78,
        place: {
          currency: "USD"
        }
      }
    ],
    usersViewing: 5,
    quantityAvailable: 10,
    percentageClaimed: 0.4
  }
};

describe(NAME, () => {
  it(`${NAME}: should render properly`, () => {
    let rendered = Renderer.create(<Component {...PROPS} />);
    expect(rendered.getInstance()).toBeDefined();
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
