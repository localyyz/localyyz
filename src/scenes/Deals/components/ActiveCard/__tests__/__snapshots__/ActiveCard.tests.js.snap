// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deals/ActiveCard Deals/ActiveCard: should render properly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "associatedPhotos": Array [
          Object {
            "imageUrl": "www.localyyz.com",
          },
        ],
        "brand": "Test",
        "place": Object {
          "currency": "USD",
        },
        "price": 6.78,
        "title": "Test Title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "marginBottom": 7.5,
          }
        }
      >
        <View
          style={
            Object {
              "padding": 15,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "backgroundColor": "rgba(255, 255, 255, 0)",
                "color": "rgb(30, 30, 30)",
                "fontSize": 22,
                "fontWeight": "700",
                "marginBottom": 7.5,
              }
            }
          >
            Tes Store
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "backgroundColor": "rgba(255, 255, 255, 0)",
                "color": "rgb(30, 30, 30)",
                "fontSize": 16,
                "fontWeight": "300",
              }
            }
          >
            Test Description
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "height": 333.5,
                  "overflow": "hidden",
                  "width": undefined,
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "opacity": 0.5,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "backgroundColor": "#efefef",
                        "borderRadius": 0,
                        "height": 333.5,
                        "width": 50,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#000000",
                  "borderRadius": 15,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "margin": 15,
                  "paddingHorizontal": 15,
                  "paddingVertical": 7.5,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "backgroundColor": "rgba(255, 255, 255, 0)",
                    "borderBottomColor": "#FFFFFF",
                    "color": "#FFFFFF",
                    "fontSize": 16,
                    "fontWeight": "600",
                  }
                }
              >
                Buy â€” $7
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "width": "100%",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "backgroundColor": "rgba(255, 255, 255, 0)",
                  "color": "#44179E",
                  "fontSize": 14,
                  "fontWeight": "600",
                  "marginBottom": 15,
                }
              }
            >
              5 currently viewing
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#50E3C2",
                  "flexDirection": "row",
                  "height": 25,
                  "justifyContent": "flex-start",
                  "width": "100%",
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "flex-end",
                    "backgroundColor": "#44179E",
                    "height": "100%",
                    "justifyContent": "center",
                    "width": "0%",
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    Object {
                      "backgroundColor": "rgba(255, 255, 255, 0)",
                      "borderBottomColor": "#FFFFFF",
                      "color": "#FFFFFF",
                      "fontSize": 12,
                      "fontWeight": "600",
                      "opacity": 0,
                      "paddingLeft": 7.5,
                      "paddingRight": 15,
                    }
                  }
                >
                  40% claimed
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "flex-start",
                    "height": "100%",
                    "justifyContent": "center",
                    "width": "100%",
                  }
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    Object {
                      "backgroundColor": "rgba(255, 255, 255, 0)",
                      "borderBottomColor": "#FFFFFF",
                      "color": "#FFFFFF",
                      "fontSize": 12,
                      "fontWeight": "600",
                      "opacity": 1,
                      "paddingLeft": 15,
                      "paddingRight": 7.5,
                    }
                  }
                >
                  40% claimed
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
